### PR TITLE
feat(router): added incoming dispute webhooks flow

### PR DESCRIPTION
**Pull request by saiharsha-juspay**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Added support for incoming dispute webhooks.

### Additional Changes

- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Manually Tested
Webhook Request
image

DB Entry
image

Outgoing webhook to merchant
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/types/transformers.rs [hunk 1]
```diff
@@ -435,3 +451,77 @@ impl ForeignFrom<storage_enums::AttemptStatus> for api_enums::AttemptStatus {
...
+}
+
+impl ForeignTryFrom<api_models::webhooks::IncomingWebhookEvent> for storage_enums::DisputeStatus {
+    type Error = errors::ValidationError;
+
+    fn foreign_try_from(
+        value: api_models::webhooks::IncomingWebhookEvent,
+    ) -> Result<Self, Self::Error> {
+        match value {
+            api_models::webhooks::IncomingWebhookEvent::DisputeOpened => Ok(Self::DisputeOpened),
```
Comment (jarnura) [inline crates/router/src/types/transformers.rs:486]: why can't we use frunk here?

File: migrations/2023-03-15-185959_add_dispute_table/up.sql [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    updated_at VARCHAR(255),
+    created_at TIMESTAMP NOT NULL DEFAULT now()::TIMESTAMP,
+    modified_at TIMESTAMP NOT NULL DEFAULT now()::TIMESTAMP
+);
+
+ALTER TYPE "EventClass" ADD VALUE 'disputes';
+
+ALTER TYPE "EventObjectType" ADD VALUE 'dispute_details';
+
+ALTER TYPE "EventType" ADD VALUE 'dispute_opened';
```
Comment (jarnura) [inline migrations/2023-03-15-185959_add_dispute_table/up.sql:29]: can we merge these all value by a single query

File: crates/router/src/core/webhooks.rs [hunk 3]
```diff
@@ -187,6 +188,165 @@ async fn refunds_incoming_webhook_flow<W: api::OutgoingWebhookType>(
...
+                id,
+                merchant_account.storage_scheme,
+            )
+            .await
+            .change_context(errors::WebhooksFlowError::ResourceNotFound),
+        _ => Err(errors::WebhooksFlowError::ResourceNotFound).into_report(),
+    }
+}
+
+async fn get_updated_dispute_object(
```
Comment (jarnura) [inline crates/router/src/core/webhooks.rs:226]: this is `get_or_update_dispute_object`

File: migrations/2023-03-15-185959_add_dispute_table/up.sql [hunk 4]
```diff
@@ -0,0 +1,35 @@
+CREATE TYPE "DisputeStage" AS ENUM ('pre_dispute', 'dispute', 'pre_arbitration');
+
+CREATE TYPE "DisputeStatus" AS ENUM ('dispute_opened', 'dispute_expired', 'dispute_accepted', 'dispute_cancelled', 'dispute_challenged', 'dispute_won', 'dispute_lost');
+
```
Comment (jarnura) [inline migrations/2023-03-15-185959_add_dispute_table/up.sql:4]: Create INDEX for queryable fields

File: crates/router/src/core/webhooks.rs [hunk 5]
```diff
@@ -187,6 +188,165 @@ async fn refunds_incoming_webhook_flow<W: api::OutgoingWebhookType>(
...
+            .get_dispute_details(request_details)
+            .change_context(errors::WebhooksFlowError::WebhookEventObjectCreationFailed)?;
+        let payment_attempt = get_payment_attempt_from_object_reference_id(
+            state.clone(),
+            webhook_details.object_reference_id,
+            &merchant_account,
+        )
+        .await?;
+        let option_dispute = db
+            .find_by_payment_id_connector_dispute_id(
```
Comment (jarnura) [inline crates/router/src/core/webhooks.rs:306]: When we create the dispute in our system? only through `webhooks`?

File: crates/storage_models/src/dispute.rs [hunk 6]
```diff
@@ -1 +1,102 @@
...
+#[diesel(table_name = dispute)]
+pub struct Dispute {
+    #[serde(skip_serializing)]
+    pub id: i32,
+    pub dispute_id: String,
+    pub amount: String,
+    pub currency: String,
+    pub dispute_stage: storage_enums::DisputeStage,
+    pub dispute_status: storage_enums::DisputeStatus,
+    pub payment_id: String,
```
Comment (jarnura) [inline crates/storage_models/src/dispute.rs:38]: How the dispute is linked to our storage model, the relation-ship model of dispute.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/webhooks.rs (hunk 5 above)
Comment (saiharsha-juspay) [reply crates/router/src/core/webhooks.rs:306]: Yes, only throgh webhooks

**Review by vspecky — DISMISSED** [review]
LGTM

File: crates/router/src/types/transformers.rs (hunk 1 above)
Comment (saiharsha-juspay) [reply crates/router/src/types/transformers.rs:486]: DisputeStatus constain subset of values of IncomingWebhookEvent. There are values in IncomingWebhookEvent which are not in DisputeStatus.

File: crates/storage_models/src/dispute.rs (hunk 6 above)
Comment (saiharsha-juspay) [reply crates/storage_models/src/dispute.rs:38]: A payment_attempt can have multiple disputes. So it's 1-N (1 to many) relationship for payment_attempt with dispute.

File: migrations/2023-03-15-185959_add_dispute_table/up.sql (hunk 4 above)
Comment (saiharsha-juspay) [reply migrations/2023-03-15-185959_add_dispute_table/up.sql:4]: Both DisputeStage and DisputeStatus are not queryable fields.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/api_models/src/disputes.rs [hunk 7]
```diff
@@ -1 +1,22 @@
+use masking::Serialize;
 
+use crate::enums::{DisputeStage, DisputeStatus};
+
+#[derive(Default, Clone, Debug, Serialize)]
+pub struct DisputeResponse {
```
Comment (jarnura) [inline crates/api_models/src/disputes.rs:6]: Add Documentation via utopia

Comment (jarnura) [reply crates/api_models/src/disputes.rs:6]: Can we done in a separate PR too.

**Review by jarnura — CHANGES_REQUESTED** [review]
Remove absolute path imports

File: crates/router/src/core/utils.rs [hunk 8]
```diff
@@ -149,3 +151,65 @@ mod tests {
...
+    };
+    if !(dispute_stage_validation && dispute_status_validation) {
+        super::metrics::INCOMING_DISPUTE_WEBHOOK_VALIDATION_FAILURE_METRIC.add(
+            &super::metrics::CONTEXT,
+            1,
+            &[],
+        );
+        Err(errors::WebhooksFlowError::DisputeWebhookValidationFailed)?
+    }
+    Ok(())
```
Comment (jarnura) [inline crates/router/src/core/utils.rs:214]:
```suggestion
    if !(dispute_stage_validation && dispute_status_validation) {
        super::metrics::INCOMING_DISPUTE_WEBHOOK_VALIDATION_FAILURE_METRIC.add(
            &super::metrics::CONTEXT,
            1,
            &[],
        );
        Err(errors::WebhooksFlowError::DisputeWebhookValidationFailed)?
    }
    Ok(())
    when !(dispute_stage_validation && dispute_status_validation) {
        super::metrics::INCOMING_DISPUTE_WEBHOOK_VALIDATION_FAILURE_METRIC.add(
            &super::metrics::CONTEXT,
            1,
            &[],
        );
        Err(errors::WebhooksFlowError::DisputeWebhookValidationFailed)?
    }
```

File: crates/storage_models/src/query/dispute.rs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            dsl::merchant_id
+                .eq(merchant_id.to_owned())
+                .and(dsl::payment_id.eq(payment_id.to_owned()))
+                .and(dsl::connector_dispute_id.eq(connector_dispute_id.to_owned())),
+        )
+        .await
+    }
+
+    #[instrument(skip(conn))]
+    pub async fn update(self, conn: &PgPooledConn, dispute: DisputeUpdate) -> StorageResult<Self> {
```
Comment (jarnura) [inline crates/storage_models/src/query/dispute.rs:38]: why merchant_id is required `dispute_id` already has unique indexing on it.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/utils.rs [hunk 10]
```diff
@@ -149,3 +151,65 @@ mod tests {
...
+}
+
+pub fn validate_dispute_stage_and_dispute_status(
+    prev_dispute_stage: DisputeStage,
+    prev_dispute_status: DisputeStatus,
+    dispute_stage: DisputeStage,
+    dispute_status: DisputeStatus,
+) -> CustomResult<(), errors::WebhooksFlowError> {
+    let dispute_stage_validation = validate_dispute_stage(&prev_dispute_stage, &dispute_stage);
+    let dispute_status_validation = if dispute_stage == prev_dispute_stage {
```
[the remark this answers is not in the document]
Comment (saiharsha-juspay) [reply crates/router/src/core/utils.rs:201]: dispute status can go to any state when it's different dispute stage, where as it has some restrictions if it's switching in same dispute stage

**Review by jarnura — APPROVED** [review]

**Review by vspecky — DISMISSED** [review]
LGTM

**Review by vspecky — APPROVED** [review]
LGTM

**Review by jarnura — APPROVED** [review]

**Pull request merged**